extensions-contrib/momentsketch/src/main/java/org/apache/druid/query/aggregation/momentsketch/aggregator/MomentSketchAggregatorFactory.java

```diff
@@ -0,0 +1,279 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.momentsketch.aggregator;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.query.aggregation.Aggregator;
+import org.apache.druid.query.aggregation.AggregatorFactory;
+import org.apache.druid.query.aggregation.AggregatorFactoryNotMergeableException;
+import org.apache.druid.query.aggregation.BufferAggregator;
+import org.apache.druid.query.aggregation.momentsketch.MomentSketchWrapper;
+import org.apache.druid.query.cache.CacheKeyBuilder;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.commons.codec.Charsets;
+import org.apache.commons.codec.binary.Base64;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+public class MomentSketchAggregatorFactory extends AggregatorFactory
+{
+  public static final int DEFAULT_K = 11;
+  public static final boolean DEFAULT_COMPRESS = true;
+
+  private final String name;
+  private final String fieldName;
+  private final int k;
+  private final boolean compress;
+  private final byte cacheTypeId;
+
+  private static final byte MOMENTS_SKETCH_CACHE_ID = 0x51;
+
+  public static final String TYPE_NAME = "momentSketch";
+
+  @JsonCreator
+  public MomentSketchAggregatorFactory(
+      @JsonProperty("name") final String name,
+      @JsonProperty("fieldName") final String fieldName,
+      @JsonProperty("k") final Integer k,
+      @JsonProperty("compress") final Boolean compress
+  )
+  {
+    this(name, fieldName, k, compress, MOMENTS_SKETCH_CACHE_ID);
+  }
+
+  MomentSketchAggregatorFactory(
+      final String name,
+      final String fieldName,
+      final Integer k,
+      final Boolean compress,
+      final byte cacheTypeId
+  )
+  {
+    if (name == null) {
+      throw new IAE("Must have a valid, non-null aggregator name");
+    }
+    this.name = name;
+    if (fieldName == null) {
+      throw new IAE("Parameter fieldName must be specified");
+    }
+    this.fieldName = fieldName;
+    this.k = k == null ? DEFAULT_K : k;
+    this.compress = compress == null ? DEFAULT_COMPRESS : compress;
+    this.cacheTypeId = cacheTypeId;
+  }
+
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    return new CacheKeyBuilder(
+        cacheTypeId
+    ).appendString(name).appendString(fieldName).appendInt(k).build();
```

Comment:
the output name of the aggregator factory (`name`) doesn't need to be in the cache key

Reply:
I think "compress" should be in the cache key, since that affects the query results